src/crypto/index.js

```diff
@@ -1426,6 +1544,233 @@ Crypto.prototype._onRoomKeyEvent = function(event) {
+    } else {
+        const request = {
+            event: event,
+            methods: methods,
+            beginKeyVerification: (method) => {
+                const verifier = this.beginKeyVerification(
+                    method,
+                    sender,
+                    content.from_device,
+                    content.transaction_id,
+                );
+                this._verificationTransactions[sender][content.transaction_id].verifier
+                    = verifier;
+                return verifier;
+            },
+            cancel: () => {
+                this._baseApis.sendToDevice("m.key.verification.cancel", {
+                    [sender]: {
+                        [content.from_device]: newUserCancelledError({
+                            transaction_id: content.transaction_id,
+                        }).getContent(),
+                    },
+                });
+            },
+        };
+        this._verificationTransactions[sender][content.transaction_id] = {
+            request: request,
+        };
+        this._baseApis.emit("crypto.verification.request", request);
+    }
+};
+
+Crypto.prototype._onKeyVerificationStart = function(event) {
+    const sender = event.getSender();
+    const content = event.getContent();
+    const transactionId = content.transaction_id;
+    const deviceId = content.from_device;
+    if (!transactionId || !deviceId) {
+        // invalid request, and we don't have enough information to send a
+        // cancellation, so just ignore it
+        return;
+    }
+    let handler = this._verificationTransactions[sender]
+          && this._verificationTransactions[sender][transactionId];
+    if (!(content.method in this._verificationMethods)) {
+        const err = newUnknownMethodError({
+            transaction_id: content.transactionId,
+        });
+        if (handler.verifier) {
+            handler.verifier.cancel(err);
+        } else if (handler.request) {
+            handler.request.cancel(err);
+        }
+        this.sendToDevice(
+            "m.key.verification.cancel", {
+                [sender]: {
+                    [deviceId]: err.getContent(),
+                },
+            },
+        );
+        return;
+    } else if (content.next_method) {
+        if (!(content.next_method in this._verificationMethods)) {
+            const err = newUnknownMethodError({
+                transaction_id: content.transactionId,
+            });
+            if (handler.verifier) {
+                handler.verifier.cancel(err);
+            } else if (handler.request) {
+                handler.request.cancel(err);
+            }
+            this.sendToDevice(
+                "m.key.verification.cancel", {
+                    [sender]: {
+                        [deviceId]: err.getContent(),
+                    },
+                },
+            );
+            return;
+        } else {
+            /* TODO:
```

Comment:
should this be to-done?

Reply:
The `TwoPartVerification` class isn't done yet, and it may end up being more complicated than this, so not quite done yet.